# Исследование надежности заемщиков


## Описание проекта

На основе данных о платёжеспособности клиентов исследовать влияет ли семейное положение и количество детей клиента на факт возврата кредита в срок.

**Задачи**:
1. Знакомство с данными
2. Провести предобработку данных (определить и обработать дубликаты, пропуски, заменить типы данных)
3. Провести каотегоризацию и декомпозировать датасет на 3 части в зависимости от проверямого параметра, который влияет на возврат кредита

**Датасет** - `data.csv`. Строка - клиент банка.

- `children` - количество детей
- `days_employed` - стаж работы в днях
- `dob_years` - возраст клиента
- `education` - образование
- `education_id` - образование (код уровня образования)
- `family_status` - семейное положение
- `family_status_id` - id семейного положения
- `gender` - пол
- `income_type` - статус дохода
- `debt` -  должник (булев тип)
- `total_income` - доход
- `purpose` - цель кредита

## Ознакомление с данными

In [1]:
# импорт библиотек
import pandas as pd

# загрузим датасет в data
data = pd.read_csv('/datasets/data.csv')

In [2]:
# посмотрим на первые 20 строк
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

In [3]:
# посмотрим на данные
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

Выведем количество пропущенных значений для каждого столбца.

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце обработаем на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработаем значения в этом столбце: заменим все отрицательные значения положительными с помощью метода `abs()`.

In [6]:
data['days_employed'] = data['days_employed'].abs()

Для каждого типа занятости выведем медианное значение трудового стажа `days_employed` в днях.

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставим их как есть.

Выведем перечень уникальных значений столбца `children`.

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

В столбце `children` есть два аномальных значения. Удалим строки, в которых встречаются такие аномальные значения из датафрейма `data`.

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

Ещё раз выведем перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

Заполним пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

Убедимся, что все пропуски заполнены.

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

Заменим вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

Обработаем неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведем их к нижнему регистру.

In [14]:
data['education'] = data['education'].str.lower()

Выведем на экран количество строк-дубликатов в данных. Если такие строки присутствуют - удалим их.

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

На основании диапазонов, указанных ниже, создадим в датафрейме `data` столбец `total_income_category` с категориями:

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


Например, кредитополучателю с доходом 25000 назначим категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используем собственную функцию с именем `categorize_income()` и метод `apply()`.

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

Выведем на экран перечень уникальных целей взятия кредита из столбца `purpose`.

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Создадим функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.

Используем собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучим данные в столбце `purpose` и определим, какие подстроки помогут вам правильно определить категорию.

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуем параметры, влияющие на возврат кредита

#### 3.1 Зависимость между количеством детей и возвратом кредита в срок

Для определения наличия зависимости необходимо сравнить долю должников в общем количестве клиентов по целевым группам. Посмотрим на данные в разрезе категорий, сформированных по количеству детей у клиента.

In [22]:
# Сгруппируем исходные данные из data по столбцу 'children' и применим метод .agg() с параметрами count и sum к столбцу 'debt'
# с добавлением новых столбцов 'customers_amount_total' (общее кол-во клиентов) и 'debt_amount' (количество должников).
# Результат сохраним в переменную data_debt
data_debt = data.groupby(['children']).agg(customers_amount_total=('debt', 'count'), debt_amount=('debt', 'sum'))

# Добавим в таблицу data_debt столбец 'debtors_percent' с долей должников в общем количестве клиентов каждой целевой группы
# в процентном измерении. Для этого разделим столбец 'debt_amount' на 'customers_amount_total'. Результат деления умножим
# на 100 и округлим функцией round() до 2 знаков после запятой.
data_debt['debtors_percent'] = round((data_debt['debt_amount'] / data_debt['customers_amount_total']) * 100, 2)

# Выведем получившуюся таблицу на на экран
data_debt

customers_amount_total  debt_amount  debtors_percent
children                                                      
0                          14091         1063             7.54
1                           4808          444             9.23
2                           2052          194             9.45
3                            330           27             8.18
4                             41            4             9.76
5                              9            0             0.00

Напишем функцию `get_pivot()`, которая принимает в качестве аргумента наименование категории клиентов `category_name`, а возвращает сводную таблицу с общим количеством клиентов, количеством должников и доли должников по каждой группе в заданной категории клиентов

In [23]:
def get_pivot(category_name):
    
    # Сгруппируем данные из data по колчеству детей клиентов с помощью сводной таблицы:
    # - строки: уникальные значения из столбца 'children'
    # - столбцы: 'Всего кредитополучателей', 'Всего должников', 'Доля должников'
    # - значения вычисляются по столбцу 'debt' с агрегирующими функциями: 'count', 'sum', 'mean'
    
    data_pivot = data.pivot_table(index=category_name, values='debt', aggfunc=['count', 'sum', 'mean'])

    # Переименуем столбцы в полученной сводной таблице
    data_pivot.columns = ['all_credits', 'all_debtors','share_debtors']

    # Результат выведем на экран
    return display(data_pivot)

# Выведем сводную таблицу с помощью функции get_pivot() с параметром 'children'
get_pivot('children')

all_credits  all_debtors  share_debtors
children                                         
0               14091         1063       0.075438
1                4808          444       0.092346
2                2052          194       0.094542
3                 330           27       0.081818
4                  41            4       0.097561
5                   9            0       0.000000

**Вывод** 

Наблюдается зависимость между <u>наличием</u> детей у клиентов банка и возвратом кредита в срок: доля клиентов, у которых нет детей, чаще возвращают кредит в срок, чем те клиенты, у которых есть дети.

#### 3.2 Зависимость между семейным положением и возвратом кредита в срок

Для определения наличия зависимости посмотрим на данные в разрезе категорий, сформированных по семейному статусу клиента.

In [25]:
# Выведем сводную таблицу с помощью функции get_pivot() с параметром 'family_status'
get_pivot('family_status')

all_credits  all_debtors  share_debtors
family_status                                                 
Не женат / не замужем         2796          273       0.097639
в разводе                     1189           84       0.070648
вдовец / вдова                 951           63       0.066246
гражданский брак              4134          385       0.093130
женат / замужем              12261          927       0.075606

**Вывод** 

Существует зависимость между семейным положением клиента банка и возвратом кредита: клиенты, которые состоят в гражданском браке или не женатые (не замужем), более склоны не возвращать кредиты в срок, чем клиенты, состоящие или состоявшие в зарегистрированном браке.

#### 3.3 Зависимость между уровнем дохода и возвратом кредита в срок

In [27]:
# Выведем сводную таблицу с помощью функции get_pivot() с параметром 'total_income_category'
get_pivot('total_income_category')

all_credits  all_debtors  share_debtors
total_income_category                                         
A                               25            2       0.080000
B                             5014          354       0.070602
C                            15921         1353       0.084982
D                              349           21       0.060172
E                               22            2       0.090909

**Вывод** 

Отследить зависимость невозможно. По данным результата анализа можно утверждать, что клиенты, доход которых находится выше среднего (B) и ниже среднего (D), являются более ответственными в части возврата кредитов в срок, чем клиенты, находящиеся в крайних (A, E) и средней (C) группах.

#### 3.4 Влияние цели кредита на его возврат в срок

In [29]:
# Выведем сводную таблицу с помощью функции get_pivot() с параметром 'purpose_category'
get_pivot('purpose_category')

all_credits  all_debtors  share_debtors
purpose_category                                                 
операции с автомобилем           4279          400       0.093480
операции с недвижимостью        10751          780       0.072551
получение образования            3988          369       0.092528
проведение свадьбы               2313          183       0.079118

**Вывод** 

Клиенты, которые берут кредит на операции с автомобилем или на получение образования менее склоны возвращать кредит в срок, чем клиенты, получающие кредит на операции с недвижимостью или на проведение свадьбы.

### Шаг 4: Общий вывод

1. На вход был дан .csv файл с данными о платежеспособности клиентов с 21525 записями. В нем содержатся данные по количеству детей клиента `children`, по семейному положению клиента `family_status`, по его ежемесячному доходу `total_income`, о наличии просрочки `debt`, а также данные о целях получения кредита `purpose`
2. На этапе предобработки данных выяснилось, что в столбцах `days_employed`, `total_income` содержится одинаковое количество пропущенных значений (2174 шт.). Данные пропуски заполнили медианным значением ежемесячного дохода по каждому типу источника дохода `income_type`
3. Далее заменили тип данных в столбце `total_income` с вещественного на целочисленный
4. В столбце `education` находились неявные дубликаты. Привели значения к нижнему регистру
5. Далее нашли явные дубликаты и избавились от них
6. Для удобства анализа клиентов по уровню дохода написали функцию `categorize_income()`, которая присваивает категорию клиенту по его ежемесячному заработку (A-E: A - клиенты с высоким доходом, E - клиенты с низким доходом). Категория отражена в столбце `total_income_category` в таблице `data`
7. Выведя уникальные значения из столбца `purpose`, обнаружилось, что в данных есть повторяющиеся значения по смыслу. Написали функцию `categorize_purpose()` по объединению целей получения кредитов к единым формулировкам
8. Далее был проведен анализ зависимости возврата кредитов от количества детей, семейного положения, уровнем дохода и целей займа клиента. Используя функцию `get_pivot()` мы получали сводную таблицу, в которой содержится информация о доле должников в каждой исследуемой категории клиентов

Таким образом, результаты анализа показали, что семейное положение и количество детей клиента влияет на факт погашения кредита в срок:
1. Клиенты, не состоящие в официальном браке, чаще не возвращают кредит в срок, чем те клиенты, которые состоят или состояли в браке;
2. Клиенты, у которых нет детей, чаще возвращают кредиты в срок, чем те клиенты, у которых есть дети.